### PR TITLE
Episode 1

### PR DESCRIPTION
**Pull request by colinsauze**
Here's my first draft of material for episode 1. 

I'm still not totally convinced that we should cover the entire deep learning pipeline here. It feels like a bit too much cognitive load and its not fully used in this episode. Would it be so bad to gradually introduce these concepts as we encounter them?

### CODE REVIEW

**Comment by colinsauze** [comment]
Tagging @noramcgregor to review this as (a very late) part of the carpentries lesson development study groups.

File: _episodes/01-introduction.md [hunk 1]
```diff
@@ -3,13 +3,247 @@ title: "Introduction"
...
+
+* Where only small amounts of training data is available.
+* Tasks requiring an explanation of how the answer was arrived at.
+* Being asked to classify things which are nothing like their training data.
+
+### What sort of problems can deep learning solve, but shouldn't be used for?
+
+Deep learning needs a lot of computational power, for this reason it often relies on specialist hardware like graphical processing units (GPUs). Many computational problems can be solved using less intensive techniques, but could still technically be solved with deep learning.
+
+The following could technically be achieved using deep learning, but it would probably be a very wasteful way to do it:
```
Comment (dafnevk) [inline _episodes/01-introduction.md:72]:
```suggestion
The following could technically be achieved using deep learning, but it would probably be very wasteful way to do it:
```

File: _episodes/01-introduction.md [hunk 2]
```diff
@@ -3,13 +3,247 @@ title: "Introduction"
...
+> 3. Identifying MRI images of a rare disease when only one or two example images available for training.
+> 4. Identifying people in pictures after being trained only on cats and dogs.
+> 5. Translating English into French.
+>
+> > ## Solution
+> >
+> > 1.  and 5 are the sort of tasks often solved with deep learning.
+> > 2. is technically possible but deep learning is massively overkill, you could do the same with much less computing power using traditional techniques.
+> > 3. will probably fail because there's not enough training data.
+> > 4. will fail because the deep learning system only knows what cats and dogs look like, it might accidentally classify the people as cats or dogs.
```
Comment (dafnevk) [inline _episodes/01-introduction.md:92]: Maybe add here (or in the explanation above) that it's a bit more nuanced: sometimes we can do transfer learning to finetune a network that was changed on a different, but similar task.

File: _episodes/01-introduction.md [hunk 3]
```diff
@@ -3,13 +3,247 @@ title: "Introduction"
...
+
+The loss function tells the training algorithm how far away the predicted value was from the true value. There are a number of different types of function we can use here that might be suited to different types of data and network architectures. We'll look at choosing a loss function in more detail later on.
+
+
+### 6. Train the model
+
+Before we can train the model we have one final component to decide upon, the optimizer. The optimizer is responsible for taking the output of the loss function and then applying some changes to weights within the network. We can think of the training process as like trying to gradually descend a (very jagged looking) curve on a graph (that looks more like a rocky mountain side than a smooth curve) and find the bottom of it, without knowing in advance where that bottom is. Along the way we might hit some flat or even rising parts of the curve that make it look like we've reached the bottom when we haven't, we call these local minima. Different optimizers and optimizer parameters can help prevent us getting stuck in local minima. As with the loss function the best choice may depend on the kind of data and network architecture that we're using.
```
Comment (dafnevk) [inline _episodes/01-introduction.md:135]: This is a great explanation, but I think it would help to add some visuals here. Ideally a gif that shows the optimization process, jumping around the loss landscape

File: _episodes/01-introduction.md [hunk 4]
```diff
@@ -3,13 +3,247 @@ title: "Introduction"
...
+Hyperparameters are all the parameters set by the person configuring the machine learning instead of those learned by the algorithm itself. The Hyperparameters include the number of epochs or the parameters for the optimizer. If we are really trying to optimize things then we might even do a sweep through a whole range of Hyperparameters and use the loss function to decide which Hyperparameters performed best.
+
+### 8. Measure Performance
+
+Once we think the network is performing well we want to measure its performance. To do this we might use the validation set of data that we put aside earlier and didn't use as part of the training process. For a classification task we can often evaluate four different measures for each possible class in the data:
+
+1. True Positives: Where we correctly classified something as belonging to this class.
+2. True Negatives: Where we correctly classified that something wasn't a member of this class.
+3. False Positives: Where we incorrectly classified something as belonging to this class when it doesn't.
+4. False Negatives: Where we incorrectly classified something as not belonging to this class when it does.
```
Comment (dafnevk) [inline _episodes/01-introduction.md:152]: I would leave this out, it is fairly specific for classification and is easier to explain with an example. Maybe just mention in general that there are several ways of computing one score that tells us how well our model performs (such as accuracy or F1 for classification, RMSE for regression) and that the choice of this metric depends on the context of the model

File: _episodes/01-introduction.md [hunk 5]
```diff
@@ -3,13 +3,247 @@ title: "Introduction"
...
+> > You should get a version number reported. At the time of writing 2.4.0 is the latest version.
+> > ~~~
+> > 2.4.0
+> > ~~~
+> > {:.output}
+> {:.solution}
+{:.challenge}
+
+> ## Testing Seaborn Installation
+> Lets check you have a suitable version of Keras installed.
```
Comment (dafnevk) [inline _episodes/01-introduction.md:214]:
```suggestion
> Lets check you have a suitable version of seaborn installed.
```

File: _episodes/01-introduction.md [hunk 6]
```diff
@@ -3,13 +3,247 @@ title: "Introduction"
...
+> > ~~~
+> > 0.11.1
+> > ~~~
+> > {:.output}
+> {:.solution}
+{:.challenge}
+
+
+> ## Testing Sklearn Installation
+> Lets check you have a suitable version of Keras installed.
```
Comment (dafnevk) [inline _episodes/01-introduction.md:232]:
```suggestion
> Lets check you have a suitable version of sklearn installed.
```

**Review by dafnevk — COMMENTED** [review]
I think this episode looks quite good, the explanation of neural networks and deep learning read good to me (but I think it would be especially beneficial to let a novice in deep learning review this episode).

I think the explanation of a neuron, as well as the explanation of optimization, could benefit from an additional illustration.

This episode is probably quite suited for self-directed learning as it is, but I'm curious how it will work out in a teaching setting. It is currently a lot of text so maybe we need to add some tips to the instructor notes on how to make this part more interactive. Visualizations could help, maybe also adding some small questions or exercises.

File: _episodes/01-introduction.md (hunk 4 above)
Comment (colinsauze) [reply _episodes/01-introduction.md:152]: I think we should probably just say something like there are many ways to achieve this and we'll discuss the details of them later.

**Comment by colinsauze** [comment]
@dafnevk 
> 
> This episode is probably quite suited for self-directed learning as it is, but I'm curious how it will work out in a teaching setting. It is currently a lot of text so maybe we need to add some tips to the instructor notes on how to make this part more interactive. Visualizations could help, maybe also adding some small questions or exercises.

I think most carpentries lessons start with a few paragraphs of explanatory text giving some background. Its not expected that the instructor will say exactly what is written but that they will cover the key points. I do find it easier to do this when i've got some properly written text instead of just bullet points.

I will try and add some more visualisations to it. I did have a comment to myself asking if we should have a graph illustrating the optimisation process, as you've suggested the same thing it seems like a good idea.

File: _episodes/01-introduction.md [hunk 7]
```diff
@@ -3,13 +3,241 @@ title: "Introduction"
...
+> Discuss your answers with the group or the person next to you.
+{: .challenge}
+
+## Deep Learning Libraries
+
+There are many software libraries available for deep learning including:
+
+### TensorFlow
+
+[TensorFlow](https://www.tensorflow.org/) was developed by Google and is one of the older deep learning libraries, first released to the public in 2015. It was originally developed for the C++ programming language and doesn't always follow a "pythonic" style that might make it seem a bit different to other libraries in Python. Its been ported to many other languages including Java, JavaScript, Ruby and Swift. It works on a very low level of individual "tensors" or arrays. Its capable of much more than deep learning and offers a very versatile toolkit for tensor operations. As a result it often takes a lot more lines of code to write deep learning operations in TensorFlow than other libraries. It offers (almost) seamless integration with GPU accelerators and Google's own TPU (Tensor Processing Unit) chips that are built specially for machine learning.
```
Comment (tobyhodges) [inline _episodes/01-introduction.md:172]:
```suggestion
[TensorFlow](https://www.tensorflow.org/) was developed by Google and is one of the older deep learning libraries, first released to the public in 2015. It was originally developed for the C++ programming language and doesn't always follow a "pythonic" style that might make it seem a bit different to other libraries in Python. It has been ported to many other languages including Java, JavaScript, Ruby and Swift. It works on a very low level of individual "tensors" or arrays. It is capable of much more than deep learning and offers a very versatile toolkit for tensor operations. As a result it often takes a lot more lines of code to write deep learning operations in TensorFlow than other libraries. It offers (almost) seamless integration with GPU accelerators and Google's own TPU (Tensor Processing Unit) chips that are built specially for machine learning.
```

Comment (tobyhodges) [reply _episodes/01-introduction.md:172]: I found the detail about "tensors or arrays" confusing here. How important is it to talk this much about TensorFlow if you will not use it for the rest of your lesson? E.g. could you cut this down to

> [TensorFlow](https://www.tensorflow.org/) was developed by Google and is one of the older deep learning libraries, ported across many languages since it was first released to the public in 2015. It is very versatile and capable of much more than deep learning but as a result it often takes a lot more lines of code to write deep learning operations in TensorFlow than in other libraries. It offers (almost) seamless integration with GPU accelerators and Google's own TPU (Tensor Processing Unit) chips that are built specially for machine learning.

**Review by tobyhodges — COMMENTED** [review]
This is a very nice theoretical overview of deep learning. I found the explanations easy to follow and I definitely learned a thing or two as I read through. Thanks for taking the time to write this!

I have made some specific corrections and suggestions, and I have three more general comments too.

1. You variously use "Deep Learning," "Deep learning," and "deep learning" which may be distracting for learners and instructors working through this material. I recommend settling on one way of writing this for the whole lesson: it will save you time at the proofreading/polishing stage.
2. To help motivate learners and make the lesson feel more directly applicable, I think it would be good to include some examples of where DL has been successfully applied to real world (research?) applications.
3. The information here is all relevant and very well explained but I think the lesson will work better if you can get into the practical aspects as quickly as possible. Specifically, I think you can cut down the "Deep learning workflow" section: you do not need to go into this much detail yet - save it for when the learners can follow along with you and this context is most relevant to the next step. you could probably replace this with a very brief bullet point list and a flow diagram and move the blocks of text to individual sections later in the lesson. you have effectively written excellent introductory paragraphs for the next nine episodes of the lesson!
I admit that this might be incompatible with your Deep learning workflow discussion exercise, but I think you could split that up into multiple smaller discussions throughout the lesson as well.

File: _episodes/01-introduction.md [hunk 8]
```diff
@@ -3,13 +3,241 @@ title: "Introduction"
...
+
+Now that we've set the optimizer function we can go ahead and start training our network. We'll probably keep doing this for a given number of iterations or epochs or until the loss function gives a value under a certain threshold. The graphs below shows the loss (generally) reducing as the number of epochs increases. Notice that on a few occasions it goes up instead of down. In the first graph it appears to have reached a plateau of near zero by the 500th epoch. The second graph shows the 500th to 1500th epoch and we can see that its still going down, but frequently rises a little before dropping again.
```
[the remark this answers is not in the document]
Comment (colinsauze) [reply _episodes/01-introduction.md:137]: My point was that even after 500 epochs its still improving, but looking at the graph of the first 500 epochs you might not realise that as the line looks flat. Whether 500 epochs is good enough really depends on what you're trying to achieve, what level of accuracy you want and whether you are just over-fitting the training data. I was thinking the discussion on how good is good enough is one to have later on after we look at measuring performance in detail.

[the remark this answers is not in the document]
Comment (colinsauze) [reply _episodes/01-introduction.md:137]: @tobyhodges what does _epochs_ do?

[the remark this answers is not in the document]
Comment (tobyhodges) [reply _episodes/01-introduction.md:137]: > @tobyhodges what does _epochs_ do?

It seemed to be the first time the word is used in the lesson and it seems to be common practice for key terms, especially those that are likely to be unfamiliar to the learner, to be emphasised in some way e.g. by italics. This is also where you could link to its definition in a glossary of terms.

File: _episodes/01-introduction.md [hunk 9]
```diff
@@ -3,13 +3,241 @@ title: "Introduction"
...
+- "When is it successful?"
+- "What are the tools involved?"
+- "What is the workflow for deep learning?"
+- "Why we chose Keras"
 objectives:
-- "First learning objective. (FIXME)"
+- "Recall the sort of problems for which Deep learning is a useful tool"
+- "List some of the available tools for Deep Learning"
+- "Recall the steps of a Deep Learning workflow"
+- "Understand why it is important to test the accuracy of deep learning system."
```
[the remark this answers is not in the document]
Comment (colinsauze) [reply _episodes/01-introduction.md:16]: Does "explain" imply we feature this in an exercise (we don't currently in this episode).

File: _episodes/01-introduction.md (hunk 7 above)
Comment (colinsauze) [reply _episodes/01-introduction.md]: resolved in commit 3b868d7

File: _episodes/01-introduction.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (tobyhodges) [reply _episodes/01-introduction.md:16]: Good question. The switch from "understand" to "explain" is primarily to make this learning objective (a little) more specific and more measurable. 
While I recommend that you have a formative assessment to address every desired outcome from your lesson/episodes, it is not a hard rule. I think you can view it as an acknowledgment that it would be nice to have a way to assess this, but you are not obligated to create that assessment right now. Another way of thinking about this is that it might be an outcome from this introductory episode that is only assessed later in the lesson e.g. when you come to the episode that discusses model accuracy in more detail.

**Review by dafnevk — APPROVED** [review]
Looks good to me, I think this episode can be merged.

**Comment by colinsauze** [comment]
I wanted to reduce the amount of text in the deep learning workflow section first.

**Comment by dafnevk** [comment]
Ah sorry Colin, then I was a bit too quick. Best to create a new PR for that, then?
